First commit for PANOS

GGabriele opened the pull request:
Hey there,
I'm opening this to keep track of updates during the process of adding PANOS support to NAPALM. 

At this time, this module supports the following methods:
- `open`
- `close`
- `load_merge_candidate`
- `load_replace_candidate`
- `compare_config`
- `discard_config`
- `commit_config`
- `rollback`
- `get_facts`
- `get_interfaces`

Let me know if you have any questions.

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+import pan.xapi
+import os.path
+import xml.etree
+import os
```
Why both ```os``` and ```os.path```?

GGabriele replied:
Good catch, I probably forgot about this during the development :)

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+        self.device = None
+        self.connected = False
+
+    def __enter__(self):
```
There's no need to override the methods ```__enter__```, ```__exit__``` and ```__raise_clean_exception```.

dbarrosop replied:
As @mirceaulinic pointed out, do not override methods that have code in `napalm-base` unless it's super strictly necessary. And when I mean super strictly necessary I really mean it ;)

GGabriele replied:
Roger that :)

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+        self.device = None
+        self.connected.disconnect()
+
+    def load_template(self, template_name, **template_vars):
```
Same here -- this method is already defined in the base class.

mirceaulinic replied:
And anyway we are doing some changes right now: https://github.com/napalm-automation/napalm-base/pull/27

mirceaulinic commented:
Great stuff! It would be even better if you could provide some tests too.

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+        if self.connected is False:
+            self._open_ssh()
+
+        output = self.connected.send_command("show config diff")
```
```self.connected``` isn't a flag that you use to check if the connection was successfully open and should use instead ```self.device```?

GGabriele replied:
Probably I made this tricky. Basically, I'm using PAN API most of the time to interact with the device, but I also use netmiko in some cases when needed. So, `self.device` is the reference to the API object while `self.connected` is the reference to the netmiko object. I'm going to change it and create a `self.ssh_device` to make this more clear. My bad :)

mirceaulinic commented on napalm_panos/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# the License.
+
+"""napalm_panos package."""
+from panos import PANOSDriver
```
<CR> maybe?

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+            print(epilog)
+        raise exc_type, exc_value, exc_traceback
+
+    def open(self):
```
Please raise ```napalm_base.exceptions.ConnectionException``` when connection cannot be established.

mirceaulinic replied:
See: https://github.com/napalm-automation/napalm-junos/blob/master/napalm_junos/junos.py#L52

GGabriele replied:
Sure! :)

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+            if self.device.status == 'success':
+                self.loaded = False
+            else:
+                print "Error while loading last-saved config."
```
Please avoid print -- better raise exceptions (https://github.com/napalm-automation/napalm-base/blob/master/napalm_base/exceptions.py). If you need a different one, feel free to open a PR for a new one.

mirceaulinic commented on MANIFEST.in:
```diff
@@ -0,0 +1,3 @@
+include requirements.txt
+include napalm_skeleton/templates/*.j2
```
As you don't have any templates defined yet, you can remove lines 2 & 3. But when you'll need, make sure to include using ```include naplam_panos/templates/*.j2``` etc.

mirceaulinic commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+napalm_base
+pan-python
```
I noticed you import ```netmiko```, add it in the requirements too.

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+import xml.etree
+import os
+import requests
+import requests_toolbelt
```
```requests_toolbelt``` doesn't depend on some libs that should be specified in ```requirements.txt```?

mirceaulinic commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+from napalm_base.exceptions import ReplaceConfigException, MergeConfigException
+
+from napalm_base.base import NetworkDriver
+from netmiko import ConnectHandler
```
Add ```netmiko``` in reqs.

dbarrosop commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+            if self.loaded is False:
+                if filename.endswith('.xml') is False:
+                    raise MergeConfigException('File must be in XML format.')
+                elif self._save_backup() is False:
```
Are these two `if` mutually exclusive? Shouldn't those be just regular `if`s? I understand that by the nature of the algorithm the `if...elif...` here works but they are supposed to be two independent checks, right?

GGabriele replied:
You're definitely right. My bad.

dbarrosop commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+
+    def load_replace_candidate(self, filename=None, config=None):
+        if config:
+            raise ReplaceConfigException("This method requires a config file.")
```
Can't you forge the configuration from a string? Do you always need a filename?

GGabriele replied:
The best way to do a replace is using an XML file and import it into the device...otherwise, the string may simply be a copy/paste of that file content. I see this just too intricate, but if you want I can code it ;)

dbarrosop commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,326 @@
+            self._open_ssh()
+
+        output = self.connected.send_command("show config diff")
+        print output
```
Don't do `print`, use `return` instead.

dbarrosop commented:
In addition to the inline comments:

1. Please, provide tests. Would be cool if you could help us getting a vagrant box as well.
2. Revisit all the prints in the code. If it's a result it should be a return (like `compare_config`) and if it's an error raise an exception. It's fine if we need new one but I would be surprised. Also, don't log with prints. I would recommend using logging and having a flag `DEBUG=False` that you can toggle.

For example:

https://github.com/dbarrosop/chatops_demo/blob/master/ansible/tasks/helpers/ansible_helper.py

GGabriele commented:
Thanks @mirceaulinic @dbarrosop for your feedbacks...I was exactly looking for them :)
I've fixed all the things you noted. Let me know if you find anything else.

@dbarrosop Regarding tests, I was already working on them and I hope to submit them soon. As for the vagrant box, I'll let you know, I have to ask Jason.

GGabriele commented:
@dbarrosop @mirceaulinic I've added more fixes and some tests as well. I hope to add more of them tomorrow :)

mirceaulinic commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,121 @@
+        self.device.load_replace_candidate(filename='%s/initial.xml' % self.vendor)
+        diff = self.device.compare_config()
+
+        print(diff)
```
Please remove prints from the tests too.

mirceaulinic commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,121 @@
+        self.device.load_replace_candidate(filename='%s/new_good.xml' % self.vendor)
+        commit_diff = self.device.compare_config()
+
+        print(commit_diff)
```
This print statement should not be here either.

mirceaulinic commented:
@GGabriele It already looks better! Except of a couple of small adjustments I mentioned in the comments, I would add:
- the tests you wrote in ```TestConfigPANOSDriver``` look very good. It would be even better if you could also define methods for the cases when it might fail (i.e.: ```test_merge_wrong_configuration``` when you try to load a wrong config etc.) and prove that you raise a specific exception in that case (using ```assertRaises```).
- import ```TestGettersNetworkDriver``` from ```napalm_base``` and inherit it to define a tester for the getters (e.g.: https://github.com/napalm-automation/napalm-junos/blob/master/test/unit/TestJunOSDriver.py#L37) and provide mock data.

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,121 @@
+    """Group of tests that test Configuration related methods."""
+
+    @classmethod
+    def setUpClass(cls):
```
You just need this method, the rest should be inherited. You can add extra if you want but don't override the existing tests. Check `napalm-ios`.

GGabriele replied:
Yeah, but I did it because `.conf` files don't work for this module.

mirceaulinic replied:
@dbarrosop I think we can keep the tests he wrote since he's not using an underneath library with such specific tests (as pyIOSXR etc)

dbarrosop replied:
I would rather look for alternatives. If we need to change the internals of the tests for some reason it's going to be a pain.

@GGabriele would it be reasonable to treat `.conf` extensions in the same way as `.xml`? In fact, you only support `XML` config right? You could ignore checking the extension of the file and raise an error when trying to parse the file with XML. The xml parser should be able to tell you if the file is a valid XML file or not (it might still be file with the `xml` extension but with complete garbage in it).

GGabriele replied:
It makes sense. Okay then, I'll work on this.

GGabriele commented:
@dbarrosop I made the changes you requeste. Let me know what you think :) 

@mirceaulinic I've also added the test you were talking about. Tomorrow I hope to add those for Getters as well :)

mirceaulinic commented on .travis.yml:
```diff
@@ -0,0 +1,10 @@
+- git clone https://github.com/napalm-automation/napalm-base.git napalm-automation/napalm-base
+- cd napalm-automation/napalm-base
+- python setup.py install
+- cd ../..
```
Can you the following lines, in order to execute the tests as well?

```yaml
script:
- cd test/unit
- nosetests -v TestPANOSDriver:TestGetterPANOSDriver.test_get_facts
- nosetests -v TestPANOSDriver:TestGetterPANOSDriver.test_get_interfaces
- cd ../..
```

GGabriele replied:
Sure :)

mirceaulinic replied:
Probably @dbarrosop will update the```deploy``` section later.

dbarrosop replied:
Yes, I can take care about the deployment phase myself. Don't worry about
that.
On Mon 9 May 2016 at 14:48, Mircea Ulinic <notifications@github.com> wrote:

> In .travis.yml
> <https://github.com/napalm-automation/napalm-panos/pull/1#discussion_r62491921>
> :
>
> > @@ -0,0 +1,10 @@
> > +language: python
> > +python:
> > +- 2.7
> > +install:
> > +- pip install -r requirements.txt
> > +- pip install .
> > +- git clone https://github.com/napalm-automation/napalm-base.git napalm-automation/napalm-base
> > +- cd napalm-automation/napalm-base
> > +- python setup.py install
> > +- cd ../..
>
> Probably @dbarrosop <https://github.com/dbarrosop> will update thedeploy
> section later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/napalm-automation/napalm-panos/pull/1/files/9836e79a0471a43cd38a298f0ad44f45448ec4d6..6908a1f0638373fbbb41452a30a036a648e20847#r62491921>
>

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,109 @@
+        cls.device.load_replace_candidate(filename='%s/initial.conf' % cls.vendor)
+        cls.device.commit_config()
+
+    def test_custom_replacing_config_and_diff_and_discard(self):
```
Can you add a brief docstring explaining why/what you are testing this? Just a couple of lines so it's easier to understand the reasoning.

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,109 @@
+        result = (commit_diff.strip() == intended_diff.strip()) and (discard_diff == '')
+        self.assertTrue(result)
+
+    def test_custom_merge_configuration(self):
```
Can you add a brief docstring explaining why/what you are testing this? Just a couple of lines so it's easier to understand the reasoning.

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,109 @@
+            cls.device.open()
+
+
+class FakePANOSDevice:
```
Brilliant : )

dbarrosop commented:
@GGabriele when you fix the few inline comments that we have I will merge the code. Then it would be great if you could help me getting access to a panos device so I can test. We might have to iterate and do minor adjustments before the first release but overall looks very good right now.

Btw, we will have to update the documentation as well before the release:

https://github.com/napalm-automation/napalm

Excellent job.

GGabriele commented:
Thanks @dbarrosop :D 

I've updated the PR as requested. I've also changed the `load_merge_candidate` logic a little bit and added one more custom test :)

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,141 @@
+# Copyright 2016 Dravetech AB. All rights reserved.
```
call this file `TestDriver`. I want to standardize the names.

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,141 @@
+from napalm_base.test.base import TestConfigNetworkDriver, TestGettersNetworkDriver
+
+
+class TestConfigPANOSDriver(unittest.TestCase, TestConfigNetworkDriver):
```
Call this class `TestConfigDriver`

dbarrosop commented on test/unit/TestPANOSDriver.py:
```diff
@@ -0,0 +1,141 @@
+        self.assertTrue(result)
+
+
+class TestGetterPANOSDriver(unittest.TestCase, TestGettersNetworkDriver):
```
And this one `TestGetterDriver`

dbarrosop commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+napalm-base
+pan-python
+netmiko
```
Please, state minimum version.

dbarrosop commented on .travis.yml:
```diff
@@ -0,0 +1,15 @@
+- cd ../..
+script:
+- cd test/unit
+- nosetests -v TestPANOSDriver:TestGetterPANOSDriver.test_get_facts
```
Please, update these to reflect the new names.

dbarrosop commented on napalm_panos/panos.py:
```diff
@@ -0,0 +1,357 @@
+        self.ssh_device.send_config_set(config)
+        self.loaded = True
+
+    def load_merge_candidate(self, filename=None, file_format='set', config=None, from_xpath=None, to_xpath=None, mode=None):
```
This method has extra parameters not supported by the API.
